webapp/src/idbworkspace.ts

```diff
@@ -18,116 +18,134 @@ const TEXTS_TABLE = "texts";
+    if (_dbPromise) {
+        return await _dbPromise;
     }
 
-    _db = new pxt.BrowserUtils.IDBWrapper("__pxt_idb_workspace", 1, (ev, r) => {
-        const db = r.result as IDBDatabase;
-        db.createObjectStore(TEXTS_TABLE, { keyPath: KEYPATH });
-        db.createObjectStore(HEADERS_TABLE, { keyPath: KEYPATH });
-    });
+    _dbPromise = createDbAsync();
 
-    return _db.openAsync()
-        .catch((e) => {
+    return _dbPromise;
+
+    async function createDbAsync(): Promise<pxt.BrowserUtils.IDBWrapper> {
+        const idbDb = new pxt.BrowserUtils.IDBWrapper("__pxt_idb_workspace", 1, (ev, r) => {
+            const db = r.result as IDBDatabase;
+            db.createObjectStore(TEXTS_TABLE, { keyPath: KEYPATH });
+            db.createObjectStore(HEADERS_TABLE, { keyPath: KEYPATH });
+        }, async () => {
+            await pxt.BrowserUtils.clearTranslationDbAsync();
+            await pxt.BrowserUtils.clearTutorialInfoDbAsync();
+        });
+
+        try {
+            await idbDb.openAsync();
+        } catch (e) {
             pxt.reportException(e);
             return Promise.reject(e);
-        })
-        .then(() => _db);
+        }
+
+        return idbDb;
+    }
 }
 
-function listAsync(): Promise<pxt.workspace.Header[]> {
-    return migrateBrowserWorkspaceAsync()
-        .then (() => {
-            return getDbAsync();
-        })
-        .then((db) => {
-            return db.getAllAsync<pxt.workspace.Header>(HEADERS_TABLE);
-        });
+async function listAsync(): Promise<pxt.workspace.Header[]> {
+    await migrateBrowserWorkspaceAsync();
+    const db = await getDbAsync();
+    return db.getAllAsync<pxt.workspace.Header>(HEADERS_TABLE);
 }
 
-function getAsync(h: Header): Promise<pxt.workspace.File> {
-    return getDbAsync()
-        .then((db) => {
-            return db.getAsync<StoredText>(TEXTS_TABLE, h.id);
-        })
-        .then((res) => {
-            return Promise.resolve({
-                header: h,
-                text: res.files,
-                version: res._rev
-            });
-        });
+async function getAsync(h: Header): Promise<pxt.workspace.File> {
+    const db = await getDbAsync();
+    const res = await db.getAsync<StoredText>(TEXTS_TABLE, h.id);
+    return {
+        header: h,
+        text: res.files,
+        version: res._rev
+    };
 }
 
-function setAsync(h: Header, prevVer: any, text?: ScriptText): Promise<pxt.workspace.Version> {
-    return getDbAsync()
-        .then((db) => {
-            const dataToStore: StoredText = {
-                id: h.id,
-                files: text,
-                _rev: prevVer
-            };
-            return (text ? db.setAsync(TEXTS_TABLE, dataToStore) : Promise.resolve())
-                .then(() => {
-                    return db.setAsync(HEADERS_TABLE, h);
-                });
-        });
+async function setAsync(h: Header, prevVer: any, text?: ScriptText): Promise<void> {
```

Comment:
Worth a note on the return type change; the return type was void / nothing was actually getting returned by `db.setAsync` before so no actual difference, it was just passing as written because `pxt.workspace.Version` is `any` and that allows void returns